Runtime.ImportModuleError: Error: Cannot find module '@pulumi/pulumi/runtime/resource.js

### What happened?

Hi,

We're having error below in lambda function after we deployed to AWS.
```
2023-11-27T19:04:19.772Z	undefined	ERROR	Uncaught Exception 	{
    "errorType": "Runtime.ImportModuleError",
    "errorMessage": "Error: Cannot find module '@pulumi/pulumi/runtime/resource.js'\nRequire stack:\n- /var/task/__index.js\n- /var/runtime/index.mjs",
    "stack": [
        "Runtime.ImportModuleError: Error: Cannot find module '@pulumi/pulumi/runtime/resource.js'",
        "Require stack:",
        "- /var/task/__index.js",
        "- /var/runtime/index.mjs",
        "    at _loadUserApp (file:///var/runtime/index.mjs:1087:17)",
        "    at async Object.UserFunction.js.module.exports.load (file:///var/runtime/index.mjs:1119:21)",
        "    at async start (file:///var/runtime/index.mjs:1282:23)",
        "    at async file:///var/runtime/index.mjs:1288:1"
    ]
}
```

The module installed from local.
![image](https://github.com/pulumi/pulumi/assets/70257872/522abc26-53dd-4ad3-99cd-f68d8cf239f6)

### Example

We tried the package **@pulumi/pulumi v3.77.1 to later** and this error is happening due to this module @pulumi/pulumi/runtime/resource.js

![image](https://github.com/pulumi/pulumi/assets/70257872/5c4b77fc-fcd8-4149-8453-8b97e5e042a5)

While from **@pulumi/pulumi v3.74.0**, we don't get the error as the module is not there 
![image](https://github.com/pulumi/pulumi/assets/70257872/63d8a190-7b88-4ba3-b708-211379f4bb5e)

Please help to check on how to resolve this for higher version.

Thanks.

### Output of `pulumi about`

![image](https://github.com/pulumi/pulumi/assets/70257872/80d4a401-cc2c-4058-a8c3-ed6ce414e9ab)


### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20240912--sdk-nodejs--fix-closure-serialization-when-using-the-pulumi-sdk-inside-anonymous-functions.yaml
+++ b/changelog/pending/20240912--sdk-nodejs--fix-closure-serialization-when-using-the-pulumi-sdk-inside-anonymous-functions.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: sdk/nodejs
+  description: Fix closure serialization when using the Pulumi SDK inside anonymous functions

--- a/sdk/nodejs/resource.ts
+++ b/sdk/nodejs/resource.ts
@@ -18,7 +18,6 @@ import * as log from "./log";
 import { Input, Inputs, interpolate, Output, output } from "./output";
 import {
     getResource,
-    pkgFromType,
     readResource,
     registerResource,
     registerResourceOutputs,
@@ -1499,4 +1498,17 @@ export function parseResourceReference(ref: string): [string, string] {
     const urn = ref.slice(0, lastSep);
     const id = ref.slice(lastSep + 2);
     return [urn, id];
+}
+
+/**
+ * Extracts the package from the type token of the form "pkg:module:member".
+ *
+ * @internal
+ */
+export function pkgFromType(type: string): string | undefined {
+    const parts = type.split(":");
+    if (parts.length === 3) {
+        return parts[0];
+    }
+    return undefined;
 }

--- a/sdk/nodejs/runtime/resource.ts
+++ b/sdk/nodejs/runtime/resource.ts
@@ -29,6 +29,7 @@ import {
     CustomResourceOptions,
     expandProviders,
     ID,
+    pkgFromType,
     ProviderResource,
     Resource,
     ResourceOptions,
@@ -1277,17 +1278,4 @@ function runAsyncResourceOp(label: string, callback: () => Promise<void>, serial
             log.debug(`Resource RPC serialization requested: ${label} is behind ${resourceChainLabel}`);
         }
     }
-}
-
-/**
- * Extracts the package from the type token of the form "pkg:module:member".
- *
- * @internal
- */
-export function pkgFromType(type: string): string | undefined {
-    const parts = type.split(":");
-    if (parts.length === 3) {
-        return parts[0];
-    }
-    return undefined;
 }

